ractor/src/port/output/mod.rs

```diff
@@ -152,3 +152,150 @@ impl OutputPortSubscription {
+///     ) -> Result<Self::State, ActorProcessingErr> {
+///         Ok(State {
+///             output_port: OutputPort::default(),
+///         })
+///     }
+///
+///     async fn handle(
+///         &self,
+///         _myself: ActorRef<Self::Msg>,
+///         message: Self::Msg,
+///         state: &mut Self::State,
+///     ) -> Result<(), ActorProcessingErr> {
+///         match message {
+///             PublisherMessage::Subscribe(subscriber) => {
+///                 // Subscribes the `OutputPortSubscriber` wrapped actor to the `OutputPort`
+///                 subscriber.subscribe_to_port(&state.output_port);
+///             }
+///             PublisherMessage::Publish(value) => {
+///                 // Broadcasts the `u8` value to all subscribed actors, which will handle the type conversion
+///                 state.output_port.send(value);
+///             }
+///         }
+///         Ok(())
+///     }
+/// }
+///
+/// // The subscriber's message type demonstrates how to transform the publisher's
+/// // message type by implementing `From<T>`:
+///
+/// #[derive(Debug)]
+/// enum SubscriberMessage {
+///     Handle(String), // Subscriber's intent for message handling
+/// }
+///
+/// impl From<u8> for SubscriberMessage {
+///     fn from(value: u8) -> Self {
+///         SubscriberMessage::Handle(value.to_string()) // Converts u8 to String
+///     }
+/// }
+///
+/// // To subscribe a subscriber actor to the publisher and broadcast a message:
+/// struct Subscriber;
+/// #[cfg_attr(feature = "async-trait", ractor::async_trait)]
+/// impl Actor for Subscriber {
+///     type State = ();
+///     type Msg = SubscriberMessage;
+///     type Arguments = ();
+///
+///     async fn pre_start(
+///         &self,
+///         _myself: ActorRef<Self::Msg>,
+///         _: (),
+///     ) -> Result<Self::State, ActorProcessingErr> {
+///         Ok(())
+///     }
+///
+///     async fn handle(
+///         &self,
+///         _myself: ActorRef<Self::Msg>,
+///         message: Self::Msg,
+///         _state: &mut Self::State,
+///     ) -> Result<(), ActorProcessingErr> {
+///         Ok(())
+///     }
+/// }
+/// async fn example() {
+///     let (publisher_actor_ref, publisher_actor_handle) =
+///         Actor::spawn(None, Publisher, ()).await.unwrap();
+///     let (subscriber_actor_ref, subscriber_actor_handle) =
+///         Actor::spawn(None, Subscriber, ()).await.unwrap();
+///
+///     publisher_actor_ref
+///         .send_message(PublisherMessage::Subscribe(Box::new(subscriber_actor_ref)))
+///         .unwrap();
+///
+///     // Broadcasting a message to all subscribers
+///     publisher_actor_ref
+///         .send_message(PublisherMessage::Publish(123))
+///         .unwrap();
+///
+///     publisher_actor_handle.await.unwrap();
+///     subscriber_actor_handle.await.unwrap();
+/// }
+/// ```
+
+#[cfg(not(feature = "cluster"))]
+pub type OutputPortSubscriber<InputMessage> = Box<dyn OutputPortSubscriberTrait<InputMessage>>;
```

Comment:
I guess the boxing provides us a way to support multiple destination message types instead of requiring a single `ActorRef<I>` type. I'm trying to think if there's a way to avoid the boxing, but I'm not sure there is yet...

Maybe a wrapper type which can contain an `ActorRef<T> where T: Into<I>` or something. The other option is taking something like an anonymous function which is `Fn(I) -> ()` which will cast to the underlying actor for example.

Just some ideas, open to any thoughts!

Reply:
I had though about this and always coming back to the need to something that needed to be sized, but I'll check your ideas soon, I may have missed something

Reply:
I'm not saying this is bad by any means but if we can avoid the vtable lookup it'd be preferable!

Anyways appreciate looking into it.

Reply:
I looked into static dispatch again but didn't find any suitable alternatives. Seems like I'm stuck with boxing the trait, which isn't what I wanted, but I don't see another option.

Using Box<dyn Trait> isn't the best for performance, but for sending subscription messages that happen infrequently, it should be fine. These messages are mainly for setting up connections between actors, so the small hit on performance seem secondary.

Then, I thought about using the registry for actor lookup. It feels like it might add more overhead because it's basically dynamic dispatch at the library level, so at first, it seems even worse than Box<dyn Trait>. But considering these messages are rare and mostly for configuration, the additional overhead might not be a significant issue. Plus, I haven't really dived into the cluster aspects of ractor yet, so there might be benefits there that outweigh these concerns.